gh-70039: smtplib: store the server name in ._host in .connect()

nmartensen opened the pull request:
Original patch by gigaplastik, extended with a few more tests.

Addresses:
<!--
- <https://github.com/python/cpython/issues/70039> / BPO 25852
- <https://github.com/python/cpython/issues/80275> / BPO 36094
-->

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-70039 -->
* Issue: gh-70039
<!-- /gh-issue-number -->
* Issue: gh-80275

nmartensen committed:
smtplib: store the server name in ._host in .connect()

Original patch by gigaplastik, extended with a few more tests.

Addresses:
https://github.com/python/cpython/issues/70039
BPO 25852

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

nmartensen committed:
move and rename tests

bitdancer commented on Lib/test/test_smtplib.py:
```diff
@@ -282,6 +282,34 @@ def testBasic(self):
+                            timeout=support.LOOPBACK_TIMEOUT)
+        # explicit .connect with host:port notation must set ._host, too
+        smtp.connect(f"{HOST}:{self.port}")
+        self.assertEqual(smtp._host, HOST)
```
I think these tests should be dropped.  At this time, `_host` is an implementation detail, and I gather they were originally implemented because of the lack of network tests.  The bug is starttls failing when connect is used explicitly, and the smtpnet tests check that.

bitdancer commented on Misc/NEWS.d/next/Library/2024-02-10-21-25-22.gh-issue-70039.6wvcAP.rst:
```diff
@@ -0,0 +1 @@
+In smtplib.SMTP, set ._host in .connect() and not immediately during instantiation. This allows .starttls() to work if the instance was created without arguments and host and port were only provided via an explicit .connect() call. If will also set ._host to the correct host value if the host and port were specified using the 'host:port' notation, which according to the documentation is supported ...
```
I would rephrase this to not talk about the implementation detail of `_host`: "Fixed bug where :meth:`SMTP.starttls` could fail if :meth:`SMTP.connection` is called explicitly rather than implicitly."  Everything else is implementation details and belongs in the commit message.

nmartensen committed:
Drop test_smtplib tests again

There are already other tests for the functionality, therefore there is
no need to test implementation details.

nmartensen committed:
Revised NEWS entry

nmartensen committed:
add missing backticks to NEWS message

nmartensen committed:
Merge branch 'main' into 70039_smtplib_host_port

nmartensen committed:
NEWS entry: add missing smtplib context
